oci/private/image.sh

```diff
@@ -1,144 +1,174 @@
+  jq -n '{
+    schemaVersion: 2, 
+    mediaType: "application/vnd.oci.image.manifest.v1+json", 
+    config: { mediaType: "application/vnd.oci.image.config.v1+json", size: 0 },
+    layers: []
+  }' | update_manifest
+  # Create the image config when there is annotations
+  jq -n --argjson platform "$platform" '{config:{}, rootfs:{type: "layers", diff_ids:[]}} + $platform' | update_config > /dev/null
 }
 
-# Upon exiting, stop the registry and print STDERR on non-zero exit code.
-trap "on_exit" EXIT
-
-# Redirect stderr to the $STDERR temp file for the rest of the script.
-exec 2>>"${STDERR}"
-
-# shellcheck disable=SC1090
-source "${REGISTRY_IMPL}" 
-REGISTRY=$(start_registry "${OUTPUT}" "${REGISTRY_STDERR}")
+function base_from() {
+  local path="$1"
+  # shellcheck disable=SC2045
+  for blob in $(ls -1 -d "$path/blobs/"*/*); do
+    local relative=${blob#"$path/"}
+    mkdir -p "$OUTPUT/$(dirname "$relative")"
+    cat "$blob" >"$OUTPUT/$relative"
+  done
+  cat "$path/oci-layout" >"$OUTPUT/oci-layout"
+  jq '.manifests[0].annotations["org.opencontainers.image.ref.name"] = "intermediate"' "$path/index.json" >"$OUTPUT/index.json"
+}
 
+function get_config() {
+  regctl blob get "$REF" "$(regctl manifest get "$REF" --format "{{.Config.Digest}}")"
+}
 
-function base_from_scratch() {
-    local raw_platform="$1"
-    local ref="${REGISTRY}/scratch:latest"
-    ref="$("${CRANE}" append --oci-empty-base -t "${ref}" "-f" "{{empty_tar}}")"
-    ref=$("${CRANE}" config "${ref}" | "${JQ}"  ".rootfs.diff_ids = [] | .history = []" | "${CRANE}" edit config "${ref}")
-    ref=$("${CRANE}" manifest "${ref}" | "${JQ}"  ".layers = []" | "${CRANE}" edit manifest "${ref}")
+function update_config() {
+  local digest=
+  local config=
+  config="$(cat -)"
+  digest="$(echo "$config" | regctl blob put "$REF")"
+  get_manifest | jq '.config.digest = $digest | .config.size = $size' --arg digest "$digest" --argjson size "${#config}" | update_manifest >/dev/null
+  echo "$digest"
+}
 
-    IFS='/' read -r -a platform <<< "$raw_platform"
+function get_manifest() {
+  regctl manifest get "$REF" --format "raw"
+}
 
-    local filter=".os = \"${platform[0]}\" | .architecture = \"${platform[1]}\""
+function update_manifest() {
+  regctl manifest put "$REF"
+}
 
-    if [ -n "${platform[2]:-}" ]; then
-        filter+=" | .variant = \"${platform[2]}\""
+function add_layer() {
+  local path=
+  local media_type=
+  local compression=
+  local digest=
+  local diffid=
+  local size=
+
+  path=$(realpath "$1")  # TODO: might not be necessary.
+  digest=$(regctl digest <"$path")
+  diffid="$digest"
+  size=$(wc -c "$path" | awk '{print $1}')
+
+  if [[ $(coreutils od -An -t x1 --read-bytes 2 "$path") == " 1f 8b" ]]; then
+    compression="gzip"
+    diffid=$(zstd --decompress --format=gzip <"$path" | regctl digest)
+  elif zstd -t "$path" 2>/dev/null; then
+    compression="zstd"
+    diffid=$(zstd --decompress --format=zstd <"$path" | regctl digest)
```

Comment:
nit: the diffid assignments are the same if you use `$compression` and could be consolidated after the conditional?

Reply:
I like them being explicitly called, makes it easier to understand.